pkg/controller/csr/certificatesigningrequest.go

```diff
@@ -0,0 +1,225 @@
+	ctrl     controller.Controller
+}
+
+// Init initializes CSR controller
+func (c *Controller) Init(mgr manager.Manager) error {
+	if ver, err := k8s.GetVersion(); err != nil {
+		return fmt.Errorf("could not determine K8s version: %s", err)
+	} else if ver.LessThanOrEqual(minSupportedK8sVersion) {
+		logrus.Warnf("CSR auto-approve not supported on K8s version %s", ver)
+		c.ctrl = nil
+		c.client = nil
+		return nil
+	}
+
+	c.client = mgr.GetClient()
+	c.scheme = mgr.GetScheme()
+	c.recorder = mgr.GetEventRecorderFor(ControllerName)
+
+	var err error
+	// Create a new controller
+	c.ctrl, err = controller.New(ControllerName, mgr, controller.Options{Reconciler: c})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StartWatch starts the watch on CertificateSigningRequests
+func (c *Controller) StartWatch() error {
+	if c.ctrl == nil || c.client == nil {
+		logrus.Warn("CSR.StartWatch() bailing out  (not initialized)")
+		return nil
+	}
+
+	err := c.ctrl.Watch(
+		&source.Kind{Type: &certv1.CertificateSigningRequest{}},
+		&handler.EnqueueRequestForObject{},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile will automatically approve Portworx-requested certificates
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var (
+		csr certv1.CertificateSigningRequest
+		ret reconcile.Result
+		log = logrus.WithField("Request.Name", req.Name)
+	)
+
+	if !util.IsTLSEnabled() {
+		log.Debugf("CSR.Reconcile() bailing out  (TLS not enabled)")
+		return ret, nil
+	}
+
+	log.Infof("Reconciling CSR")
+
+	if err := c.client.Get(ctx, req.NamespacedName, &csr); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debugf("Ignoring deleted CSR request")
+			return ret, nil
+		}
+		log.WithError(err).Errorf("Unable to get CSR")
+		return ret, err
+	}
+
+	// ignore spec.signerName != kubernetes.io/kubelet-serving
+	if csr.Spec.SignerName != certv1.KubeletServingSignerName {
+		log.Debugf("Ignoring CSR with signer %s", csr.Spec.SignerName)
+		return ret, nil
+	}
+
+	// ignore names which don't start with "px-"
+	if !strings.HasPrefix(csr.GetName(), "px-") {
+		log.Debugf("Ignoring CSR  (not px name)")
+		return ret, nil
+	}
+
+	// ignore spec.username != system:serviceaccount:<namespace>:portworx
+	if !reMatchUserAnyNamespace.MatchString(csr.Spec.Username) {
+		log.Debugf("Ignoring CSR  (wrong user %s)", csr.Spec.Username)
+		return ret, nil
+	}
+
+	if approved, denied := c.getCertApprovalCondition(&csr.Status); approved || denied {
+		log.WithField("approved", approved).WithField("denied", denied).
+			Debugf("CSR already approved/denied")
+		return ret, nil
+	}
+
+	if cnt := len(csr.Status.Certificate); cnt > 0 {
+		log.Debugf("CSR Certificate already signed  (%d certs attached)", cnt)
+		return ret, nil
+	}
+
+	// ignore no labels, or missing label name=portworx, or missing label cid
+	var clusterID string
+	if labs := csr.GetLabels(); labs == nil {
```

Comment:
nit: We can check `len(labs) == 0`. That way we check for both `nil` and empty map.